src/GafferScene/RenderController.cpp

```diff
@@ -219,6 +222,70 @@ struct ObjectInterfaceHandle : public boost::noncopyable
 // Internal implementation details
 //////////////////////////////////////////////////////////////////////////
 
+// Maps between scene locations and integer IDs.
+class RenderController::IDMap
+{
+
+	public :
+
+		// Creates an ID if no ID has been created yet.
+		uint32_t acquireID( const ScenePlug::ScenePath &path )
+		{
+			Mutex::scoped_lock lock( m_mutex, /* write = */ false );
+			auto it = m_map.find( path );
+			if( it != m_map.end() )
+			{
+				return it->second;
+			}
+
+			lock.upgrade_to_writer();
+			return m_map.insert( PathAndID( path, m_map.size() + 1 ) ).first->second;
+		}
+
+		uint32_t idForPath( const ScenePlug::ScenePath &path ) const
+		{
+			Mutex::scoped_lock lock( m_mutex, /* write = */ false );
+			auto it = m_map.find( path );
+			if( it != m_map.end() )
+			{
+				return it->second;
+			}
+			return 0;
```

Comment:
Under what circumstances would we hit this branch?  Probably not an exception, and it might get hammered a lot, so probably better not to warn, but I generally don't like silent failures.

Reply:
At least, we should probably document in the header that 0 is a reserved value meaning no id found.

Reply:
> Under what circumstances would we hit this branch?

Any time we query a path that hasn't been rendered by the RenderController. The most common cause in practice is when converting the scene selection into IDs for drawing. When switching scenes, the currently selected paths might not exist in the new scene, or when selecting from another editor, the selected path might not be expanded by the RenderController.

> At least, we should probably document in the header that 0 is a reserved value meaning no id found.

Good point. I've squashed that into https://github.com/GafferHQ/gaffer/pull/4667/commits/f06bb979b70ba0525945bcb2a051d7a11893d717.